File: src/main/java/FlinkConsumer.java
package consumer;

import com.stefan_grafberger.streamdq.checks.RowLevelCheckResult;
import io.opencensus.trace.Link;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.kafka.clients.consumer.ConsumerConfig;

import java.util.LinkedList;
import java.util.List;
import java.util.Properties;

import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;
import com.stefan_grafberger.streamdq.VerificationSuite;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;
import org.apache.flink.configuration.Configuration;

public class FlinkConsumer {
    private static final Logger logger = LogManager.getLogger(FlinkConsumer.class);
    public static List<List<RowLevelCheckResult<TaxiRideData>>> res = new LinkedList<>();
    public static void main(String[] args) {
        try {
            logger.info("Starting Flink Consumer application...");

            StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
            logger.info("Initialized StreamExecutionEnvironment.");

            Properties props = new Properties();
            props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "kafka:9092");
            props.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "flink-consumer-group");
            logger.info("Kafka consumer properties set.");

            FlinkKafkaConsumer<String> kafkaConsumer = new FlinkKafkaConsumer<>("topic", new SimpleStringSchema(),
                    props);

            logger.info("Created Kafka Consumer.");

            env.addSource(kafkaConsumer)
                    .process(new StreamDQValidator(res));
            logger.info("Added Kafka Consumer as source to the environment.");

            env.execute("Flink Consumer with StreamDQ Validation");
            logger.info("Flink Consumer application started.");

            logger.info("~~~~~~~~~~~~~~res~~~~~~~~~~~~~~~~~~~~~~~~~");
            logger.info(res.toString());
            logger.info("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");

        } catch (Exception e) {
            logger.error("An error occurred during the execution of the Flink Consumer application", e);
        }
    }

    public static class StreamDQValidator extends ProcessFunction<String, String> {
        private transient long lastTimeWindow;
        private transient int recordCount;
        private transient long totalLatency;
        private transient int latencyRecordCount;

        public List<List<RowLevelCheckResult<TaxiRideData>>> sub_res = null;
        StreamDQValidator(List<List<RowLevelCheckResult<TaxiRideData>>> res){
            this.sub_res = new LinkedList<>(res);
        }
        @Override
        public void open(Configuration parameters) throws Exception {
            lastTimeWindow = System.currentTimeMillis();
            recordCount = 0;
            totalLatency = 0;
            latencyRecordCount = 0;
        }

        @Override
        public void processElement(String jsonString, Context ctx, Collector<String> out) {
            logger.debug("Processing element: {}", jsonString);
            long startProcessingTime = System.currentTimeMillis();
            try {
                recordCount++;
                Validation validation = new Validation(jsonString);

                // 行级检查：确保乘客计数在0到5之间
                logger.debug("Initialized validation.");

                VerificationSuite verificationSuite = new VerificationSuite();
                this.sub_res.add(validation.getRes());


                long endProcessingTime = System.currentTimeMillis();
                long latency = endProcessingTime - startProcessingTime;
                totalLatency += latency;
                latencyRecordCount++;

                long now = System.currentTimeMillis();
                if (now - lastTimeWindow >= 1000) { // 每1000毫秒（1秒）计算一次吞吐量和平均延迟
                    double averageLatency = latencyRecordCount == 0 ? 0 : (double) totalLatency / latencyRecordCount;
                    logger.info("Throughput (records per second): {}, Average Latency: {} ms", recordCount,
                            averageLatency);

                    // 重置计数器、总延迟和时间窗口
                    recordCount = 0;
                    totalLatency = 0;
                    latencyRecordCount = 0;
                    lastTimeWindow = now;
                }
                // 将数据原样输出
                out.collect(jsonString);
                logger.debug("Element processed and collected: {}", jsonString);
            } catch (Exception e) {
                logger.error("Error processing element with StreamDQValidator", e);
                throw new RuntimeException("Error processing element with StreamDQValidator", e);
            }
        }
    }
}

File: src/main/scala/FlinkConsumer.scala
package consumer

import com.stefan_grafberger.streamdq.checks.row.RowLevelCheck
import org.apache.flink.api.common.functions.RichFlatMapFunction
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer010
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.api.java.tuple.Tuple2
import org.apache.flink.util.Collector
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.logging.log4j.LogManager

import java.util.Properties

/*class FlinkConsumer extends RichFlatMapFunction[String, Tuple2[String, Integer]] {
  private val logger = LogManager.getLogger(this.getClass)
  private var inputDataStream: DataStream[String] = _
  val env = StreamExecutionEnvironment.getExecutionEnvironment

  override def open(parameters: Configuration): Unit = {
    logger.info("FlinkConsumer started.")

  }

  override def flatMap(value: String, out: Collector[Tuple2[String, Integer]]): Unit = {
    logger.info(s"Received string: $value")
    //change struct
    val cleaned_data = value.replaceAll("[{}\"]", "")
    val parts = cleaned_data.split(",")
    val eData = parts.map { part =>
      val keyValue = part.split(":")
      if (keyValue.length == 2) Some(keyValue(1)) else None
    }.flatten
    //save as taxiridedata
    val taxi_data_stream = env.fromElements(
      TaxiRideData(
        eData(0),eData(1),eData(2),eData(3),eData(4),eData(5),
        eData(6),eData(7),eData(8),eData(9),eData(10),eData(11),
        eData(12),eData(13),eData(14),eData(15),eData(16),eData(17),
        eData(18),eData(19)
      )
    )
    val valid = new Validation(env, taxi_data_stream)


    out.collect(new Tuple2(value, 1))
  }

  override def close(): Unit = {
    logger.info("FlinkConsumer closed.")
  }
}

object FlinkConsumer {
  private val logger = LogManager.getLogger(getClass)
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    val props = new Properties()
    props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "kafka:9092")
    props.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "flink-consumer-group")
    logger.info("Connecting to Kafka...")

    // Create a consumer
    val kafkaConsumer = new FlinkKafkaConsumer010[String]("topic", new SimpleStringSchema(), props)
    logger.info("Kafka consumer created for topic 'topic'.")

    // Add Kafka consumer as a source 
    val kafkaStream = env.addSource(kafkaConsumer)

    logger.info("Started consuming data from Kafka.")

    // Apply a flatMap transformation
    val qualityCheck = kafkaStream.flatMap(new FlinkConsumer)
    qualityCheck.print()
    //no key?
    //val keyedStream = kafkaStream.keyBy { ->}


    try {
      // Execute the Flink job
      env.execute("Flink Consumer")
      logger.info("Flink consumer execution started.")
    } catch {
      case e: Exception => logger.error("An error occurred while executing Flink consumer.", e)
    }
  }
}*/
